Code-gen: Remove lowercasing for project imports

jsturtevant opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
If a project has a uppercase letter in the project import path then the generated paths will end up in different directories when some of the files are lower-cased during generation.  This PR fixes the few places where the imports are lower-cased.  

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/code-generator/issues/20

**Special notes for your reviewer**:
There was another PR (#63892) that was started but closed with no explanation.  It looks like it was missing a few spots area's where the lower casing was missed which are addressed in this PR.

I was not sure where or if there are tests for this section of code.  Would be happy to add tests if you can point me in the right direction.  

If you run the `code-generator/generate-groups.sh` against the project [here](https://github.com/Azure/azure-k8s-metrics-adapter/tree/code-gen-bug) you can see the lower-casing in action.  The generated code is split across two directories and the project does not build but with these fixes code-generation completes as expected and the project builds.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Code-gen: Remove lowercasing for project imports
```

jsturtevant committed:
Remove lowercasing for project imports

This commit provides a fix for the scenario where a project has an
uppercase letter in the project import path. Prior to this fix
the generated files would end up in different directories with some
of the imports being lower-cased during generation. An example of this would
be a project such as 'github.com/MixedCase/project' would result in
some of the imports with 'github.com/mixedcase/project' causing a broken
build.

jsturtevant commented:
/assign @caesarxuchao

yue9944882 commented:
/cc @kubernetes/code-generator-maintainers

caesarxuchao commented:
/ok-to-test

jsturtevant commented:
/retest

jsturtevant commented:
@caesarxuchao I was able to pass the tests with a few updates.  

I started to add an example of the Uppercasing in _examples like the previous PR.  The generation works as expected (I haven't commited this yet).

But during that further testing I notice that the verify tests don't actually run the smoke tests in the verify script.  From the [test output](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/68484/pull-kubernetes-verify/106590/?log#log) of this PR:

```
I0913 00:03:18.965] Smoke testing _example by compiling...
I0913 00:03:19.008] Building client-gen
W0913 00:03:19.108] warning: "vendor/k8s.io/code-generator/_example/..." matched no packages
W0913 00:03:19.109] warning: "vendor/k8s.io/code-generator/_example/..." matched no packages
```

There `_example` folder looks to have been renamed to _example**s** at some point.  I made an update to build the projects by updating `staging/src/k8s.io/code-generator/hack/verify-codegen.sh` to build the samples for example:

```
go build ${SCRIPT_ROOT}/_examples/crd/...
```

but now I get the error:

```
./hack/verify-codegen.sh
Generating deepcopy funcs
Generating defaulters
Generating conversions
Generating clientset for example:v1 example2:v1 at k8s.io/code-generator/_examples/apiserver/clientset
Generating listers for example:v1 example2:v1 at k8s.io/code-generator/_examples/apiserver/listers
Generating informers for example:v1 example2:v1 at k8s.io/code-generator/_examples/apiserver/informers
Generating deepcopy funcs
Generating clientset for example:v1 example2:v1 at k8s.io/code-generator/_examples/crd/clientset
Generating listers for example:v1 example2:v1 at k8s.io/code-generator/_examples/crd/listers
Generating informers for example:v1 example2:v1 at k8s.io/code-generator/_examples/crd/informers
Generating deepcopy funcs
Generating clientset for example:v1 at k8s.io/code-generator/_examples/UpperCaseProjectName/clientset
Generating listers for example:v1 at k8s.io/code-generator/_examples/UpperCaseProjectName/listers
Generating informers for example:v1 at k8s.io/code-generator/_examples/UpperCaseProjectName/informers
diffing ./hack/../_examples against freshly generated codegen
./hack/../_examples up to date.
Smoke testing _example by compiling...
./hack/..
../../../../vendor/k8s.io/code-generator/_examples/crd/clientset/versioned/typed/example/v1/clustertesttype.go:28:2: cannot find package "k8s.io/kubernetes/pkg/apis/autoscaling" in any of:
	/home/jstur/go/src/github.com/kubernetes/kubernetes/vendor/k8s.io/kubernetes/pkg/apis/autoscaling (vendor tree)
	/usr/local/go/src/k8s.io/kubernetes/pkg/apis/autoscaling (from $GOROOT)
	/home/jstur/go/src/k8s.io/kubernetes/pkg/apis/autoscaling (from $GOPATH)
```

Since the smoke tests have not been running it seems this has been an issue before this PR as you can see on this [file in master](https://github.com/kubernetes/kubernetes/blob/6eeff3e6c94452581a94d16769ea406dec848152/staging/src/k8s.io/code-generator/_examples/crd/clientset/versioned/typed/example/v1/clustertesttype.go#L28).  Looking at the vendor folder it seems that the all those packages are symlinked back into the main `pkg` folder.  Although none of them have the prefix `kubernetes`.  It seems that there might be another bug related to how those autoscaling imports are being generated.  

Thoughts on how to proceed?  I could check in the samples folder with the Uppercase as an example, get this merged and closed, then open a separate bug for the re-enabling the smoke tests and fixing the generating of the autoscaling package?  What do you think?

jsturtevant commented:
@caesarxuchao all the tests pass is this good to merge?  I can open issues for the comments above separately.

caesarxuchao commented on staging/src/k8s.io/code-generator/cmd/client-gen/generators/client_generator.go:
```diff
@@ -130,7 +130,7 @@ func DefaultNameSystem() string {
 
 func packageForGroup(gv clientgentypes.GroupVersion, typeList []*types.Type, clientsetPackage string, groupPackageName string, groupGoName string, apiPath string, srcTreePath string, inputPackage string, boilerplate []byte) generator.Package {
-	groupVersionClientPackage := strings.ToLower(filepath.Join(clientsetPackage, "typed", groupPackageName, gv.Version.NonEmpty()))
+	groupVersionClientPackage := filepath.Join(clientsetPackage, "typed", groupPackageName, strings.ToLower(gv.Version.NonEmpty()))
```
Can you lower case the group package name as well?  Here and other places.

I don't see any reason why mixed case is needed there.

jsturtevant replied:
addressed in the latest updates.

caesarxuchao reviewed (commented):
Can you add a test demonstrating the client-gen works for camelcased path now?

jsturtevant committed:
Add MixedCase example project

jsturtevant committed:
Re-enable smoke tests in verify-codegen script

The smoke tests were not being run for the example projects.  Re-enabled
the smoke tests by building each of the sample projects.

jsturtevant committed:
Fix code generation by adding missing parameter

The code generation for fake types was missing the subresource path
parameter in the template which caused a compile error for the
sample projects using the scale subresource.  Also re-ran the code
generation after applying the fix.

jsturtevant commented:
I added an example with MixedCase. Also needed to update the verify script to build the packages correctly as stated in my [previous comment](https://github.com/kubernetes/kubernetes/pull/68484#issuecomment-421143025).  

Once the example projects were building properly,  this uncovered a bug in the code generation for the sub resource `scale` in the existing crd example:

```
Smoke testing _example by compiling...
# k8s.io/kubernetes/vendor/k8s.io/code-generator/_examples/crd/clientset/versioned/typed/example/v1/fake
vendor/k8s.io/code-generator/_examples/crd/clientset/versioned/typed/example/v1/fake/fake_clustertesttype.go:137:46: not enough arguments in call to testing.NewRootGetSubresourceAction
	have (schema.GroupVersionResource, string)
	want (schema.GroupVersionResource, string, string)
```

I fixed this in the `generator_fake_for_type.go` by adding the correct parameters.  Everything builds nows.  I also tested the updates against [my larger project](https://github.com/Azure/azure-k8s-metrics-adapter/tree/code-gen-bug) that prompted this bug fix as mentioned in the initial PR.

jsturtevant commented:
@caesarxuchao the latest commits passed the tests.  Ready for review.  Thanks!

caesarxuchao commented:
LGTM. Thanks. Can you squash it to meaningful commits?

jsturtevant commented:
@caesarxuchao squashed commits and tests passed.  Looks ready to go.  Thanks again!

caesarxuchao commented:
/lgtm

caesarxuchao commented:
/assign @lavalamp

for approval

nikhita commented:
/assign @sttts 
/kind cleanup

sttts commented on staging/src/k8s.io/code-generator/hack/verify-codegen.sh:
```diff
@@ -50,4 +50,6 @@ fi
\ No newline at end of file
+go build ./${SCRIPT_ROOT}/_examples/crd/...
+go build ./${SCRIPT_ROOT}/_examples/apiserver/...
+go build ./${SCRIPT_ROOT}/_examples/MixedCase/...
```
:+1:

sttts commented:
/lgtm
/approve
